### PR TITLE
(GH-2731) Load files from specified Puppet paths

### PR DESCRIPTION
**Pull request by lucywyman**
This introduces a new syntax for loading files from Puppet modules.
Previously, files in Puppet modules could be loaded if the file existed
in the `files/` subdirectory of a module on the modulepath, and the user
provided `mymodule/file.sh` to load a file located at
`mymodule/files/file.sh`.

The new loading behavior is gated behind a new project-level
configuration option. The option is specified under the `future` key
which is used enable different features that may introduce breaking
changes in Bolt. Right now the only suboption that `future` has is
`file_paths`, which enables the following loading behavior.

The new loading behavior loads with the following precedence depending
on what syntax is provided. Note that files can now be loaded from the
`scripts/` directory of a module in addition to the `files/` directory:

Provided: **module/myscript.sh**
- Load from `files/myscript.sh`

Provided: **module/scripts/myscript.sh**
- Load from `files/scripts/myscript.sh`
- Fall back to `scripts/myscript.sh`

Provided: **module/files/myscript.sh**
- Load from `files/files/myscript.sh`
- Fall back to `files/myscript.sh`

!feature

* **Load files from specified Puppet paths** ([#2731](https://github.com/puppetlabs/bolt/issues/2731))

  If the project-level `future.file_paths` configuration is enabled,
  Puppet files can be loaded using the new loading syntax. For more information see https://pup.pt/bolt-loading-files.

### CODE REVIEW

**Comment by lucywyman** [comment]
Just a quick note this can't go out until 7.6.0 is released later this month.

File: lib/bolt/util.rb [hunk 1]
```diff
@@ -85,6 +85,38 @@ def first_run?
         Bolt::Config.user_path && !File.exist?(first_runs_free)
       end
 
+      def find_file_in_mod(mod, module_file)
+        subdir = module_file.split(File::SEPARATOR).first
```
Comment (nicklewis) [inline lib/bolt/util.rb:89]: You can use something like `subdir, relative_path = module_file.split(File::SEPARATOR, 2)` to split the path in only one place.

File: lib/bolt/util.rb [hunk 2]
```diff
@@ -85,6 +85,38 @@ def first_run?
...
+        case subdir
+        when 'scripts'
+          script_path = module_file.split(File::SEPARATOR).drop(1).join(File::SEPARATOR)
```
Comment (nicklewis) [inline lib/bolt/util.rb:92]: `script_path` will be `""` if the original path was just `mymodule/scripts` (or `nil` if you use the `split(..., 2)` suggestion above) . In that case, we probably shouldn't fall back to `mod.script`. Ditto for the `files` case.

File: lib/bolt/util.rb [hunk 3]
```diff
@@ -85,6 +85,44 @@ def first_run?
...
+        when 'files'
+          unless relative_path.nil?
+            mod.file(module_file) || mod.file(File.join(mod, relative_path))
+          else
+            mod.file(module_file)
+          end
+        else
+          mod.file(module_file)
+        end
+      end
```
Comment (beechtom) [inline lib/bolt/util.rb:106]: This method isn't the easiest to follow. I think this stems from a couple things:

- `module_file` is always split, which results in `subdir` sometimes being the name of a subdir, while other times it ends up being the filename. Just from a naming standpoint, it seems like `subdir` should be `nil` if the path doesn't have at least two parts.
- Nested logic. `||` operators inside `unless` statements inside a `case` statement make this logically a bit hard to follow. Part of the extra complexity comes from checking `relative_path.nil?`, but I'm not sure if there's a way around that. Maybe the way to approach that is with...
- Comments. It would be helpful to have comments describing what filepath is being checked in each statement and why.

File: lib/bolt/util.rb [hunk 4]
```diff
@@ -85,6 +85,44 @@ def first_run?
         Bolt::Config.user_path && !File.exist?(first_runs_free)
       end
 
+      def find_file_in_mod(mod, module_file)
```
Comment (beechtom) [inline lib/bolt/util.rb:88]: This method, `find_module_manually`, and `search_module` should have comments describing what the method does. While they're not the most complex methods, I think it's preferable to have a couple sentences describing something instead of requiring someone to parse the code.

File: lib/bolt/util.rb [hunk 5]
```diff
@@ -85,6 +85,44 @@ def first_run?
...
+            mod.file(module_file)
+          end
+        when 'files'
+          unless relative_path.nil?
+            mod.file(module_file) || mod.file(File.join(mod, relative_path))
+          else
+            mod.file(module_file)
+          end
+        else
+          mod.file(module_file)
```
Comment (beechtom) [inline lib/bolt/util.rb:104]:
```suggestion
          mod.file(module_file) || mod.script(module_file)
```

I think this needs to fallback to `scripts`?

File: bolt-modules/boltlib/lib/puppet/functions/run_script.rb [hunk 6]
```diff
@@ -108,7 +110,13 @@ def run_script_with_description(scope, script, targets, description = nil, optio
...
     executor.report_function_call(self.class.name)
 
-    found = Puppet::Parser::Files.find_file(script, scope.compiler.environment)
+    found = if executor.future&.fetch('file_paths', false)
+              Puppet::Parser::Files.find_in_module(script, scope.compiler.environment) do |mod, module_file|
+                Bolt::Util.find_file_in_mod(mod, module_file)
+              end
+            else
+              Puppet::Parser::Files.find_file(script, scope.compiler.environment)
+            end
```
Comment (beechtom) [inline bolt-modules/boltlib/lib/puppet/functions/run_script.rb:119]: A comment describing what's happening here would be helpful.

File: bolt-modules/boltlib/lib/puppet/functions/upload_file.rb [hunk 7]
```diff
@@ -70,7 +71,13 @@ def upload_file_with_description(scope, source, destination, targets, descriptio
...
     executor.report_function_call(self.class.name)
 
-    found = Puppet::Parser::Files.find_file(source, scope.compiler.environment)
+    found = if executor.future&.fetch('file_paths', false)
+              Puppet::Parser::Files.find_in_module(source, scope.compiler.environment) do |mod, module_file|
+                Bolt::Util.find_file_in_mod(mod, module_file)
+              end
+            else
+              Puppet::Parser::Files.find_file(source, scope.compiler.environment)
+            end
```
Comment (beechtom) [inline bolt-modules/boltlib/lib/puppet/functions/upload_file.rb:80]: Ditto on the comment.

File: bolt-modules/file/lib/puppet/functions/file/exists.rb [hunk 8]
```diff
@@ -18,8 +18,16 @@
...
+    executor = Puppet.lookup(:bolt_executor)
+    executor.report_function_call(self.class.name)
+
+    found = if executor.future&.fetch('file_paths', false)
+              Puppet::Parser::Files.find_in_module(filename, scope.compiler.environment) do |mod, module_file|
+                Bolt::Util.find_file_in_mod(mod, module_file)
+              end
+            else
+              Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+            end
```
Comment (beechtom) [inline bolt-modules/file/lib/puppet/functions/file/exists.rb:30]: Ditto on the comment.

File: bolt-modules/file/lib/puppet/functions/file/read.rb [hunk 9]
```diff
@@ -17,10 +17,16 @@
...
+    executor.report_function_call(self.class.name)
 
-    found = Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+    found = if executor.future&.fetch('file_paths', false)
+              Puppet::Parser::Files.find_in_module(filename, scope.compiler.environment) do |mod, module_file|
+                Bolt::Util.find_file_in_mod(mod, module_file)
+              end
+            else
+              Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+            end
```
Comment (beechtom) [inline bolt-modules/file/lib/puppet/functions/file/read.rb:29]: Ditto on the comment.

File: bolt-modules/file/lib/puppet/functions/file/readable.rb [hunk 10]
```diff
@@ -18,9 +18,16 @@
...
+    executor.report_function_call(self.class.name)
 
-    found = Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+    found = if executor.future&.fetch('file_paths', false)
+              Puppet::Parser::Files.find_in_module(filename, scope.compiler.environment) do |mod, module_file|
+                Bolt::Util.find_file_in_mod(mod, module_file)
+              end
+            else
+              Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+            end
```
Comment (beechtom) [inline bolt-modules/file/lib/puppet/functions/file/readable.rb:30]: image

**Comment by lucywyman** [comment]
Ok, did quite a bit of refactoring so that everything ends up using the same function to search modules. I looked through Puppet's file loading to make sure it doesn't do any other magic to load from other paths. This way there's a guaranteed consistent experience between the plan function path loading and the CLI loading.

**Comment by beechtom** [comment]
- [ ] `module/script.sh`
  - [ ] Finds `module/files/script.sh`
  - [ ] Errors if `module/files/script.sh` is missing
- [ ] `modules/scripts/script.sh`
  - [ ] Finds `module/files/scripts/script.sh`
  - [ ] Falls back to `module/scripts/script.sh` with `file_paths` configured
  - [ ] Errors if both files are missing with `file_paths` configured
  - [ ] Errors if `module/files/scripts/script.sh` is missing without `file_paths` configured
- [ ] `modules/files/script.sh`
  - [ ] Finds `module/files/files/script.sh`
  - [ ] Falls back to `module/files/script.sh` with `file_paths` configured
  - [ ] Errors if both files are missing with `file_paths` configured
  - [ ] Errors if `module/files/files/script.sh` is missing without `file_paths` configured
- `run_script`, `upload_file`, and `file::*` plan functions use new behavior
- `bolt script run` and `bolt file upload` commands use new behavior

File: bolt-modules/boltlib/lib/puppet/functions/run_script.rb [hunk 11]
```diff
@@ -108,7 +110,22 @@ def run_script_with_description(scope, script, targets, description = nil, optio
...
+              # Fall back to new-style paths if enabled.
+              fallback = executor.future&.fetch('file_paths', false)
+              # Split the module_name from the pattern used to find the file
+              module_name, file_pattern = Bolt::Util.split_file_path(script)
+              # Get the absolute path to the module root from Puppet
+              mod_path = scope.compiler.environment.module(module_name)&.path
+              # Search the module for the file, optionally falling back to new-style
+              # paths if enabled.
+              Bolt::Util.find_file(mod_path, file_pattern, fallback) if mod_path
+            end
```
Comment (beechtom) [inline bolt-modules/boltlib/lib/puppet/functions/run_script.rb:127]: Could this be a method on `Bolt::Util` since it's used in multiple plan functions? I don't imagine this will need to change, but just in case it does that would reduce risk.

File: bolt-modules/file/lib/puppet/functions/file/exists.rb [hunk 12]
```diff
@@ -12,14 +12,31 @@
...
+    required_param 'String[1]', :filename
     return_type 'Boolean'
   end
 
   def exists(scope, filename)
     # Send Analytics Report
-    Puppet.lookup(:bolt_executor) {}&.report_function_call(self.class.name)
-    found = Puppet::Parser::Files.find_file(filename, scope.compiler.environment)
+    executor = Puppet.lookup(:bolt_executor)
+    executor.report_function_call(self.class.name)
```
Comment (beechtom) [inline bolt-modules/file/lib/puppet/functions/file/exists.rb:22]: Why this change? Don't functions available in apply blocks need to have the safe operator for the executor?

File: lib/bolt/cli.rb [hunk 13]
```diff
@@ -563,7 +563,8 @@ def execute(options)
             when 'command'
               executor.run_command(targets, options[:object], executor_opts)
             when 'script'
-              script_path = find_file(options[:object])
+              script_path = find_file(options[:object], executor.future&.fetch('file_paths', false))
```
Comment (beechtom) [inline lib/bolt/cli.rb:566]: We could update `Bolt::Config` to set `'future' => {}` by default so we can get rid of the safe operator.

**Review by beechtom — COMMENTED** [review]
Couple small comments, but otherwise lgtm! The refactor of the `Bolt::Util` methods makes it a lot easier to understand what's happening as well. :)

File: bolt-modules/boltlib/lib/puppet/functions/run_script.rb (hunk 11 above)
Comment (lucywyman) [reply bolt-modules/boltlib/lib/puppet/functions/run_script.rb]: Yeah, that's a good idea. I moved this to be on the executor, which feels more correct for passing scope around and will also let any other executors easily override this function.

File: lib/bolt/cli.rb (hunk 13 above)
Comment (lucywyman) [reply lib/bolt/cli.rb:566]: I thought about that, but I think at some point it might useful to know if `future` was set at all or not. We may not end up needing that, but I don't want to rule it out so early.

File: bolt-modules/file/lib/puppet/functions/file/exists.rb (hunk 12 above)
Comment (lucywyman) [reply bolt-modules/file/lib/puppet/functions/file/exists.rb]: Oh my gosh thank you for catching this! I had completely forgotten these could be called from inside apply blocks. Ended up reworking everything so that Bolt can locate these files without an executor.

File: bolt.gemspec [hunk 14]
```diff
@@ -55,7 +55,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency "net-ssh", ">= 4.0"
   spec.add_dependency "net-ssh-krb", "~> 0.5"
   spec.add_dependency "orchestrator_client", "~> 0.5"
-  spec.add_dependency "puppet", ">= 6.18.0"
+  # spec.add_dependency "puppet", ">= 6.18.0"
```
[the remark this answers is not in the document]
Comment (lucywyman) [reply bolt.gemspec]: Haha at least you know I actually forgot to touch this last week 😅

**Review by beechtom — APPROVED** [review]
